### ISSUE TITLE
Installed Twine 2.3.9 updating SugarCube corrupts story-formats.json and gets reset

### ISSUE THREAD

**Issue by cooldevo** [klembot/twinejs#710]
Summary: When I attempt to update SugarCube to 2.33.1 (also had the same happen with 2.33.0), the story-formats.json array gets corrupted and is reset back to default. It only seems to happen with SugarCube. If I leave 2.31.1 alone and add two proofing formats (illume and poof) they stay installed and work across persistent restarts of Twine. I have not tested other story formats.

My PC is Windows 10 Home x64, with OS and Programs installed on C drive (SSD) and my user profile saved on the D drive (HDD).

My first attempts [here](https://intfiction.org/t/twine-offline-story-formats-keep-resetting-to-default/46507/15), was using the 2.3.9 I downloaded on August 3rd and was a fresh install. This morning, I uninstalled Twine, downloaded a fresh copy of Twine 2.3.9 installer, and went into %appdata% and deleted the Twine folder there, just to be sure it was a fresh install.

Installation:
- Installed for anyone who uses the computer
- To the default 'C:\Program Files\Twine' folder.

After the install completed I did NOT launch Twine. I went straight into %appdata%/Twine and opened the story-formats.json. What I see is below (I inserted the line breaks to separate elements):

```
{
"0":{"id":"82f5b3c1-cc0d-4e8a-9062-f354796bce46","name":"Chapbook","url":"story-formats/chapbook-1.2.0/format.js","userAdded":false,"version":"1.1.0"},
"1":{"id":"84ba707e-71ca-4254-9d0b-a9e1d449bcf2","name":"Harlowe","url":"story-formats/harlowe-1.2.4/format.js","userAdded":false,"version":"1.2.4"},
"2":{"id":"72a97190-8a99-46fb-9690-7db1dd830368","name":"Harlowe","url":"story-formats/harlowe-2.1.0/format.js","userAdded":false,"version":"2.1.0"},
"3":{"id":"c594c161-5410-4292-9f2b-5e356c9624a7","name":"Harlowe","url":"story-formats/harlowe-3.1.0/format.js","userAdded":false,"version":"3.1.0"},
"4":{"id":"d8168c86-6677-4d89-8188-2c5ed9b7f039","name":"Paperthin","url":"story-formats/paperthin-1.0.0/format.js","userAdded":false,"version":"1.0.0"},
"5":{"id":"d1813cab-27d8-452b-953f-5e0deb34e37a","name":"Snowman","url":"story-formats/snowman-1.4.0/format.js","userAdded":false,"version":"1.4.0"},
"6":{"id":"6cc48b29-f436-43bd-b092-147a12b4fb09","name":"Snowman","url":"story-formats/snowman-2.0.2/format.js","userAdded":false,"version":"2.0.2"},
"7":{"id":"7a0f396f-4429-4a33-8642-0a29940ad294","name":"SugarCube","url":"story-formats/sugarcube-1.0.35/format.js","userAdded":false,"version":"1.0.35"}
}
8":{"id":"51378a41-abca-403b-8f9c-6475e67003c5","name":"SugarCube","url":"story-formats/sugarcube-2.31.1/format.js","userAdded":false,"version":"2.31.1"}
}

```
As you can see, the array is already messed up, with element number 8 outside of it, and closing a non-opened array.

I did not change anything in the JSON and started Twine, let it sit for about 30 seconds then exited out. What I see is below (again line breaks added to separate elements):

```
{
"0":{"id":"3cc01b33-0cd9-4275-8295-744bb9c9db1a","name":"Chapbook","url":"story-formats/chapbook-1.2.0/format.js","userAdded":false,"version":"1.1.0"},
"1":{"id":"56c7d6f5-46cc-4304-af6f-1345df85f017","name":"Harlowe","url":"story-formats/harlowe-1.2.4/format.js","userAdded":false,"version":"1.2.4"},
"2":{"id":"792d649d-344f-4854-8bc9-6eae05351912","name":"Harlowe","url":"story-formats/harlowe-2.1.0/format.js","userAdded":false,"version":"2.1.0"},
"3":{"id":"252710e3-7aa8-408e-83b4-edf782415108","name":"Harlowe","url":"story-formats/harlowe-3.1.0/format.js","userAdded":false,"version":"3.1.0"},
"4":{"id":"b28c63c3-eacd-4024-a217-432512f92e5b","name":"Paperthin","url":"story-formats/paperthin-1.0.0/format.js","userAdded":false,"version":"1.0.0"},
"5":{"id":"b7b40388-f3d8-4cd9-bbfd-4548a6e27b7d","name":"Snowman","url":"story-formats/snowman-1.4.0/format.js","userAdded":false,"version":"1.4.0"},
"6":{"id":"9237c5a3-7b99-4d9c-84d1-6ee2b884236a","name":"Snowman","url":"story-formats/snowman-2.0.2/format.js","userAdded":false,"version":"2.0.2"},
"7":{"id":"45ff2026-84ec-49d7-b794-c89f78cea684","name":"SugarCube","url":"story-formats/sugarcube-1.0.35/format.js","userAdded":false,"version":"1.0.35"},
"8":{"id":"5a4b155c-ca0d-44a2-b7cb-747c40f121c6","name":"SugarCube","url":"story-formats/sugarcube-2.31.1/format.js","userAdded":false,"version":"2.31.1"}
}
```
Which shows a much more valid array and list of elements than when it installed.

Then I 'Add Format' through the GUI to update to SugarCube 2.33.1 (again added line breaks for separating elements) and exited out of Twine:

```
{
"0":{"id":"e93fe6ef-bafd-438c-91a7-a10bb77242a4","name":"Chapbook","url":"story-formats/chapbook-1.2.0/format.js","userAdded":false,"version":"1.1.0"},
"1":{"id":"4380a1f2-12fd-453e-babd-a8751702f6f4","name":"Harlowe","url":"story-formats/harlowe-1.2.4/format.js","userAdded":false,"version":"1.2.4"},
"2":{"id":"57b28db7-b06c-4871-8623-6cba4db4eb87","name":"Harlowe","url":"story-formats/harlowe-2.1.0/format.js","userAdded":false,"version":"2.1.0"},
"3":{"id":"1afe7971-2155-4eff-a7dd-9c4564edfd88","name":"Harlowe","url":"story-formats/harlowe-3.1.0/format.js","userAdded":false,"version":"3.1.0"},
"4":{"id":"8047836f-2207-4861-829f-18c0283e0c3e","name":"Paperthin","url":"story-formats/paperthin-1.0.0/format.js","userAdded":false,"version":"1.0.0"},
"5":{"id":"6b2d2d43-8860-4c4e-a050-af42f7f9b531","name":"Snowman","url":"story-formats/snowman-1.4.0/format.js","userAdded":false,"version":"1.4.0"},
"6":{"id":"2526ad18-36ac-4c82-a715-6e3902d60316","name":"Snowman","url":"story-formats/snowman-2.0.2/format.js","userAdded":false,"version":"2.0.2"},
"7":{"id":"40ee09b5-0c9a-4689-8206-3162dfc173fa","name":"SugarCube","url":"story-formats/sugarcube-1.0.35/format.js","userAdded":false,"version":"1.0.35"},
"8":{"id":"d7e05616-b564-4775-8ee6-bcebc62ffaaf","name":"SugarCube","url":"file:///D:/Profile/Documents/Twine/StoryFormats/sugarcube-2/format.js","userAdded":true,"version":"2.33.1"}
}
```
All is looking good here so far. All files in the SugarCube downloaded zip file are co-located in that path above, which is where the Stories and Backups folder are located.

I then start Twine, ensure the default format is set to 2.33.1, fiddle around with a story or two, and exit out. What I now see is:

```
{
"0":{"id":"20ac1002-4ecf-4726-b023-de758dde58f0","name":"Chapbook","url":"story-formats/chapbook-1.2.0/format.js","userAdded":false,"version":"1.1.0"},
"1":{"id":"7fe06b0d-1b7f-4714-ae41-dfe1a5449878","name":"Harlowe","url":"story-formats/harlowe-1.2.4/format.js","userAdded":false,"version":"1.2.4"},
"2":{"id":"e77d625d-1792-4c27-85e6-e0171e993116","name":"Harlowe","url":"story-formats/harlowe-2.1.0/format.js","userAdded":false,"version":"2.1.0"},
"3":{"id":"9e28f3af-e50e-4e5a-9373-9661fc5ef02e","name":"Harlowe","url":"story-formats/harlowe-3.1.0/format.js","userAdded":false,"version":"3.1.0"},
"4":{"id":"3757e351-8d81-4643-9899-38ba3275a80e","name":"Paperthin","url":"story-formats/paperthin-1.0.0/format.js","userAdded":false,"version":"1.0.0"},
"5":{"id":"69ecc497-bfbc-4b05-9273-c0cffaa5517b","name":"Snowman","url":"story-formats/snowman-1.4.0/format.js","userAdded":false,"version":"1.4.0"},
"6":{"id":"b1cef083-0c9b-4b43-9e46-104ea40cbc2e","name":"Snowman","url":"story-formats/snowman-2.0.2/format.js","userAdded":false,"version":"2.0.2"},
"7":{"id":"6657bb83-5336-4b88-ada4-1f8f062ad5d8","name":"SugarCube","url":"story-formats/sugarcube-1.0.35/format.js","userAdded":false,"version":"1.0.35"},
"8":{"id":"a999f04b-e7b1-43ff-9e51-d1c40968ce61","name":"SugarCube","url":"file:///D:/Profile/Documents/Twine/StoryFormats/sugarcube-2/format.js","userAdded":true,"version":"2.33.1"}
}
9":{"id":"f35193eb-6b3b-43c1-9834-8a6662d7b09d","name":"SugarCube","url":"story-formats/sugarcube-2.31.1/format.js","userAdded":false,"version":"2.31.1"}
}
```
There is a new entry at the end of the file for the old SugarCube, and formatted incorrectly closing an array that wasn't opened.

The next time I start up Twine and exit I see:

```
{
"0":{"id":"86fcc03e-20c2-40de-9fb8-a74a6f903aff","name":"Chapbook","url":"story-formats/chapbook-1.2.0/format.js","userAdded":false,"version":"1.1.0"},
"1":{"id":"a53f0450-b07a-484b-becf-574fea127e03","name":"Harlowe","url":"story-formats/harlowe-1.2.4/format.js","userAdded":false,"version":"1.2.4"},
"2":{"id":"9e9380ec-fc27-4c4a-9f40-25e8897bd53f","name":"Harlowe","url":"story-formats/harlowe-2.1.0/format.js","userAdded":false,"version":"2.1.0"},
"3":{"id":"93d360f6-763e-4472-99a4-200fbe77a383","name":"Harlowe","url":"story-formats/harlowe-3.1.0/format.js","userAdded":false,"version":"3.1.0"},
"4":{"id":"4df1be3c-da4f-47b2-b2ad-d7d5c1df77a0","name":"Paperthin","url":"story-formats/paperthin-1.0.0/format.js","userAdded":false,"version":"1.0.0"},
"5":{"id":"7315c1a8-83af-49b6-9739-384110281937","name":"Snowman","url":"story-formats/snowman-1.4.0/format.js","userAdded":false,"version":"1.4.0"},
"6":{"id":"57e88cb2-4e2b-4243-a460-875e37ad4e0e","name":"Snowman","url":"story-formats/snowman-2.0.2/format.js","userAdded":false,"version":"2.0.2"},
"7":{"id":"6a4a12c8-a7c6-4359-a972-e1554f5b1c18","name":"SugarCube","url":"story-formats/sugarcube-1.0.35/format.js","userAdded":false,"version":"1.0.35"},
"8":{"id":"4f11959d-3078-4f5b-9df8-2697e40c7d1f","name":"SugarCube","url":"story-formats/sugarcube-2.31.1/format.js","userAdded":false,"version":"2.31.1"}
}
```
My updated SugarCube is now gone, and only 2.31.1 is there. This happens every time I try to update SugarCube.

On the flip-side, if I add Poof and Illume (proofing formats) they seem to work without any issue, even across persistent restarts of the application. But if I add SugarCube after those two, it corrupts the array and everything (including the proofing formats) get blown back to the default.

At @HiEv's suggestion I tried to match Chapbook version to 1.2 and nothing changed in the outcome. You can see that in the thread at the top of this page.

**Comment by cooldevo** [comment]
A temporary work-around is that on starting with a clean story-formats.json file, to add all the formats wanted. Then exit, and once ensuring the file is clean, mark it as Read Only in Windows Explorer. This looks to prevent the extra broken element from being added back in.

Of course, to make update any formats requires making it Read Write again. And then the broken element got added into the file when I started Twine.

**Comment by gbarnabic** [comment]
I'm having the same issue on my MacBook Pro macOS Catalina 10.15.6.  Added format SugarCube 2.33.1 is not retained between sessions.

**Comment by klembot** [comment]
Thanks for the information--will try to investigate this.

**Comment by mikev37** [comment]
Was having this issue when updating from Sugarcube 2.34.1 to 2.36.1 on Windows install Twine 2.3.15 
Every time I'd relaunch the application it would reset the version to Sugarcube 2.34 and break all the 2.36 dependent scripts.

Solution was the following:
Reinstalled the application for user only, instead of for anyone.
In story.js formats I nuked every other format except for 2.36.1, and set user installed to false, and set its index to 0
In prefs.js I changed the default from 2.34 to 2.36
Set both files to "Read Only"

Seems to work correctly for now, except it found the old formats somewhere? but 2.34 doesn't exist anymore and is replaced by 2.36
